src/types/any.rs

```diff
@@ -912,6 +912,15 @@ impl PyAny {
         Ok(result == 1)
     }
 
+    /// Checks whether this object is an instance of exactly type `ty` (not a subclass).
+    ///
+    /// This is equivalent to the Python expression `type(self) is ty`.
+    #[inline]
+    pub fn is_exact_instance(&self, ty: &PyType) -> bool {
```

Comment:
`is_instance` takes `PyAny` instead of `PyType`. Would that be possible here as well or would the call to `Py_IS_TYPE` be invalid then?

Reply:
`Py_IS_TYPE` requires `*mut PyTypeObject` as the second argument, so if we didn't take `&PyType` we'd need to do a type check first. (And maybe return `PyResult<bool`> as a consequence?)

Reply:
`Py_IS_TYPE` seems to be implemented inline, i.e.

```rust
#[inline]
pub unsafe fn Py_IS_TYPE(ob: *mut PyObject, tp: *mut PyTypeObject) -> c_int {
    (Py_TYPE(ob) == tp) as c_int
}

#[inline]
pub unsafe fn Py_TYPE(ob: *mut PyObject) -> *mut PyTypeObject {
    (*ob).ob_type
}
```

so we do control the type signature to a degree. It also basically uses `tp` as the right-hand side of a pointer comparison meaning that its real type is never required.

I would not suggest we change the public signature as this would be breaking with little cause. But maybe we could implement this as

```rust
pub fn is_exact_instance(&self, ty: &PyAny) -> bool {
    unsafe { (*self.as_ptr()).ob_type == ty.as_ptr() }
}
```

instead?

(IIRC, the reason for taking `PyAny` instead of `PyType` was exactly so that the caller could avoid downcasting just to check a type.)

Reply:
I see what you mean. I think we don't need to expand `Py_TYPE`, so how about using safe code?

```rust
pub fn is_exact_instance(&self, ty: &PyAny) -> bool {
    self.get_type().is(ty)
}
```

This should optimize away

Reply:
Yes, that looks much better and should optimize out. Nice!